test/topology_experimental_raft/test_topology_upgrade_stuck.py

```diff
@@ -0,0 +1,110 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+import asyncio
+import logging
+import pytest
+import time
+
+from test.pylib.manager_client import ManagerClient
+from test.pylib.rest_client import inject_error
+from test.pylib.util import wait_for_cql_and_get_hosts
+from test.topology.util import reconnect_driver, restart, enter_recovery_state, \
+        delete_raft_data_and_upgrade_state, log_run_time, wait_until_upgrade_finishes as wait_until_schema_upgrade_finishes, \
+        wait_until_topology_upgrade_finishes, delete_raft_topology_state, wait_for_cdc_generations_publishing, \
+        check_system_topology_and_cdc_generations_v3_consistency
+
+
+@pytest.mark.asyncio
+@log_run_time
+async def test_topology_upgrade_stuck(request, manager: ManagerClient):
+    # First, force the nodes to start in legacy mode due to the error injection
+    cfg = {'error_injections_at_startup': ['force_gossip_based_join']}
+
+    servers = [await manager.server_add(config=cfg) for _ in range(3)]
+    cql = manager.cql
+    assert(cql)
+
+    logging.info("Waiting until driver connects to every server")
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    logging.info("Checking the upgrade state on all nodes")
+    for host in hosts:
+        status = await manager.api.raft_topology_upgrade_status(host.address)
+        assert status == "not_upgraded"
+
+    logging.info("Enabling error injection which will cause nodes to get stuck")
+    await asyncio.gather(*(manager.api.enable_injection(s.ip_addr, "topology_coordinator_fail_to_build_state_during_upgrade", one_shot=False) for s in servers))
+
+    logging.info("Triggering upgrade to raft topology")
+    await manager.api.upgrade_to_raft_topology(hosts[0].address)
+
+    logging.info("Waiting until upgrade starts")
+    # TODO: Wait until the nodes actually get stuck? How can we check this?
```

Comment:
How about logging a message in the injection and waiting for it?

Or perhaps something is already being logged, when the exception is thrown and catched somewhere in the coordinator loop? We could then wait for it.